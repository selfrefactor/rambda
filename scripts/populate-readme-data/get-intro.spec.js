import { getIntro } from './get-intro'

test('happy', async () => {
  expect(await getIntro()).toMatchSnapshot()
})
