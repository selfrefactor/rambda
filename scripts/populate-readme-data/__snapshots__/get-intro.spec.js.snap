// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`happy 1`] = `
"
# Rambda

\`Rambda\` is faster and smaller alternative to the popular functional programming library **Ramda**. - [Documentation](https://selfrefactor.github.io/rambda/#/)

[![CircleCI](https://circleci.com/gh/selfrefactor/rambda/tree/master.svg?style=svg)](https://circleci.com/gh/selfrefactor/rambda/tree/master)
[![codecov](https://codecov.io/gh/selfrefactor/rambda/branch/master/graph/badge.svg)](https://codecov.io/gh/selfrefactor/rambda)
[![dependencies Status](https://david-dm.org/selfrefactor/rambda/status.svg)](https://david-dm.org/selfrefactor/rambda)
![Normal size](https://img.badgesize.io/selfrefactor/rambda/master/dist/rambda.js)
![Gzip size](https://img.badgesize.io/selfrefactor/rambda/master/dist/rambda.js?compression=gzip)

## Example use

\`\`\`javascript
import { compose, map, filter } from 'rambda'

const result = compose(
  map(x => x * 2),
  filter(x => x > 2)
)([1, 2, 3, 4])
// => [6, 8]
\`\`\`

You can test this example in <a href=\\"https://rambda.now.sh?const%20result%20%3D%20R.compose(%0A%20%20R.map(x%20%3D%3E%20x%20*%202)%2C%0A%20%20R.filter(x%20%3D%3E%20x%20%3E%202)%0A)(%5B1%2C%202%2C%203%2C%204%5D)%0A%0A%2F%2F%20%3D%3E%20%5B6%2C%208%5D\\">Rambda's REPL</a>

* [Install](#install)
* [Differences between Rambda and Ramda](#differences-between-rambda-and-ramda)
* [API](#api)
* [Changelog](#changelog)

## Rambda's advantages

- Tree-shaking

Currently **Rambda** is more tree-shakable than **Ramda**

- Speed

**Rambda** is generally more performant than \`Ramda\` as the [benchmarks](#benchmarks) can prove that.

- dot notation for \`R.path\` and \`R.paths\`

Standard usage of \`R.path\` is \`R.path(['a', 'b'], {a: {b: 1} })\`.

In **Rambda** you have the choice to use dot notation(which is arguably more readable):

\`\`\`
R.path('a.b', {a: {b: 1} })
\`\`\`

- comma notation for \`R.pick\` and \`R.omit\`

Similar to dot notation, but the separator is comma(\`,\`) instead of dot(\`.\`).

\`\`\`
R.pick('a,b', {a: 1 , b: 2, c: 3} })

// No space allowed between properties
\`\`\`

- Typescript included

Typescript definitions are included in the library, in comparison to **Ramda**, where you need to additionally install \`@types/ramda\`.


- More generic methods

\`Ramda\` has an overwhelming list of methods, as one could get lost putting all these methods in one's head. \`Rambda\` has smaller method counts and that could be seen as advantage.

<details>
<summary>
  Click to see the full list of Ramda methods not implemented in Rambda 
</summary>

- __
- addIndex
- ap
- aperture
- apply
- applyTo
- ascend
- binary
- bind
- call
- chain
- comparator
- composeK
- composeP
- composeWith
- construct
- constructN
- contains
- converge
- countBy
- curryN
- descend
- differenceWith
- dissocPath
- dropLastWhile
- dropRepeats
- dropRepeatsWith
- dropWhile
- empty
- eqBy
- eqProps
- evolve
- findLast
- findLastIndex
- forEachObjIndexed
- gt
- gte
- hasIn
- hasPath
- innerJoin
- insert
- insertAll
- into
- invert
- invertObj
- invoker
- juxt
- keysIn
- lift
- liftN
- lt
- lte
- mapAccum
- mapAccumRight
- mapObjIndexed
- memoizeWith
- mergeAll
- mergeDeepLeft
- mergeDeepRight
- mergeDeepWith
- mergeDeepWithKey
- mergeLeft
- mergeRight
- mergeWith
- mergeWithKey
- move
- nAry
- nthArg
- o
- objOf
- of
- once
- or
- otherwise
- pair
- partialRight
- partition
- pathEq
- pathSatisfies
- pickBy
- pipeK
- pipeP
- pipeWith
- project
- propSatisfies
- props
- reduceBy
- reduceRight
- reduceWhile
- reduced
- remove
- scan
- sequence
- sortWith
- splitAt
- splitWhen
- symmetricDifferenceWith
- takeLastWhile
- takeWhile
- andThen
- toPairsIn
- transduce
- traverse
- tryCatch
- unapply
- unary
- uncurryN
- unfold
- union
- unionWith
- uniqBy
- unless
- unnest
- until
- useWith
- valuesIn
- when
- where
- whereEq
- xprod
- zipWith
- thunkify
- default

</details>
  
## Install

- **yarn add rambda**

- For UMD usage either use \`./dist/rambda.umd.js\` or following CDN link:

\`\`\`
https://unpkg.com/rambda@CURRENT_VERSION/dist/rambda.umd.js
\`\`\`

## Differences between Rambda and Ramda

- Rambda's **type** detect async functions and unresolved \`Promises\`. The returned values are \`'Async'\` and \`'Promise'\`.

- Rambda's **type** handle \`NaN\` input, in which case it returns \`\\"NaN\\"\`.

- Rambda's **path** and **paths** accepts dot notation(\`'x.y' same as ['x','y']\`)

- Rambda's **pick** and **omit** accept comma notation(\`'x,y' same as ['x','y']\`)

- Rambda's **map**, **reject** and **forEach** can iterate over objects not only arrays.

- Rambda's **map** and **filter** pass array index as second argument when mapping over arrays.

- Rambda's **defaultTo** accept indefinite number of arguments when non curried, i.e. \`R.defaultTo(2, foo, bar, baz)\`.

- Rambda's **adjust**, **all**, **allPass**, **any**, **anyPass**, **findIndex** , **findLastIndex** and **reject** are passing index as second argument to the predicate function.

- Rambda's **startsWith/endsWith** work only with strings, instead with array and strings.

- Rambda's **equals** doesn't protect against circular structures as **Ramda.equals** does.

- Rambda's **flip** works only for functions expecting two arguments.

- Rambda's **partial** doesn't need the input arguments to be wrapped as array.

- Rambda's **filter** returns empty array with bad input(\`null\` or \`undefined\`), while Ramda throws.

- Ramda's **includes** will throw an error if input is neither \`string\` nor \`array\`, while **Rambda** version will return \`false\`.

- Ramda's **clamp** work for letters, while Rambda's method work only for numbers.

> If you need more **Ramda** methods in **Rambda**, you may either submit a \`PR\` or check the extended version of **Rambda** - [Rambdax](https://github.com/selfrefactor/rambdax). In case of the former, you may want to consult with [Rambda contribution guidelines.](CONTRIBUTING.md)

## Benchmarks

<details>

<summary>
Click to expand all benchmark results

There are methods which are benchmarked only with \`Ramda\` and \`Rambda\`(i.e. no \`Lodash\`).

Note that some of these methods, are called with and without curring. This is done in order to give more detailed performance feedback.

</summary>

method | Rambda | Ramda | Lodash
--- |--- | --- | ---
 *add* | 96.71% slower | 96.74% slower | ğŸš€ Fastest
 *adjust* | ğŸš€ Fastest | 4.57% slower | ğŸ”³
 *all* | ğŸš€ Fastest | 95.2% slower | ğŸ”³
 *allPass* | ğŸš€ Fastest | 99.24% slower | ğŸ”³
 *any* | ğŸš€ Fastest | 98.77% slower | 16.04% slower
 *anyPass* | ğŸš€ Fastest | 99.06% slower | ğŸ”³
 *append* | ğŸš€ Fastest | 82.5% slower | ğŸ”³
 *applySpec* | ğŸš€ Fastest | 67.65% slower | ğŸ”³
 *assoc* | 91.68% slower | 67.07% slower | ğŸš€ Fastest
 *clone* | ğŸš€ Fastest | 83.23% slower | 57.63% slower
 *compose* | ğŸš€ Fastest | 95.55% slower | 76.43% slower
 *converge* | 52.86% slower | ğŸš€ Fastest | ğŸ”³
 *curry* | ğŸš€ Fastest | 54.29% slower | ğŸ”³
 *curryN* | 53.6% slower | ğŸš€ Fastest | ğŸ”³
 *defaultTo* | ğŸš€ Fastest | 73.18% slower | ğŸ”³
 *drop* | ğŸš€ Fastest | 97.91% slower | ğŸ”³
 *dropLast* | ğŸš€ Fastest | 97.64% slower | ğŸ”³
 *equals* | 79.8% slower | 82.41% slower | ğŸš€ Fastest
 *filter* | 85.25% slower | 93.45% slower | ğŸš€ Fastest
 *find* | ğŸš€ Fastest | 92.33% slower | 39.13% slower
 *findIndex* | ğŸš€ Fastest | 98.67% slower | 82.94% slower
 *flatten* | 96.51% slower | 95.67% slower | ğŸš€ Fastest
 *ifElse* | ğŸš€ Fastest | 66.96% slower | ğŸ”³
 *includes* | ğŸš€ Fastest | 73.35% slower | ğŸ”³
 *indexOf* | ğŸš€ Fastest | 86.49% slower | 0.52% slower
 *init* | 94.77% slower | 97.85% slower | ğŸš€ Fastest
 *is* | ğŸš€ Fastest | 11.92% slower | ğŸ”³
 *isEmpty* | 65.35% slower | 94.04% slower | ğŸš€ Fastest
 *last* | ğŸš€ Fastest | 99.76% slower | 2.57% slower
 *lastIndexOf* | ğŸš€ Fastest | 48.6% slower | ğŸ”³
 *map* | 50.45% slower | 75.05% slower | ğŸš€ Fastest
 *match* | ğŸš€ Fastest | 47.13% slower | ğŸ”³
 *merge* | 62.61% slower | ğŸš€ Fastest | 46.89% slower
 *none* | ğŸš€ Fastest | 89.88% slower | ğŸ”³
 *omit* | ğŸš€ Fastest | 73.66% slower | 98.13% slower
 *over* | ğŸš€ Fastest | 46.97% slower | ğŸ”³
 *path* | 5.49% slower | 78.37% slower | ğŸš€ Fastest
 *pick* | ğŸš€ Fastest | 24.75% slower | 88.3% slower
 *prop* | ğŸš€ Fastest | 89.61% slower | ğŸ”³
 *propEq* | ğŸš€ Fastest | 93.1% slower | ğŸ”³
 *range* | 95.75% slower | 92.61% slower | ğŸš€ Fastest
 *reduce* | 70.51% slower | 82.97% slower | ğŸš€ Fastest
 *repeat* | 82.83% slower | 94.99% slower | ğŸš€ Fastest
 *replace* | ğŸš€ Fastest | 31.81% slower | 1.04% slower
 *set* | ğŸš€ Fastest | 33.95% slower | ğŸ”³
 *sort* | ğŸš€ Fastest | 56.02% slower | ğŸ”³
 *sortBy* | ğŸš€ Fastest | 55.12% slower | 86.04% slower
 *split* | ğŸš€ Fastest | 85.43% slower | 29.34% slower
 *splitEvery* | ğŸš€ Fastest | 90.33% slower | ğŸ”³
 *take* | 92.07% slower | 97.97% slower | ğŸš€ Fastest
 *takeLast* | 92.69% slower | 98.86% slower | ğŸš€ Fastest
 *test* | ğŸš€ Fastest | 95.29% slower | ğŸ”³
 *type* | 35.33% slower | ğŸš€ Fastest | ğŸ”³
 *uniq* | 99.13% slower | 96.24% slower | ğŸš€ Fastest
 *update* | ğŸš€ Fastest | 85.13% slower | ğŸ”³
 *view* | ğŸš€ Fastest | 72.31% slower | ğŸ”³

</details>

## Used by

- [WatermelonDB](https://github.com/Nozbe/WatermelonDB)

- [SAP's Cloud SDK](https://github.com/SAP/cloud-sdk)

- [VSCode Slack intergration](https://github.com/verydanny/vcslack)

- [Webpack PostCSS](https://github.com/sectsect/webpack-postcss)

- [MobX-State-Tree decorators](https://github.com/farwayer/mst-decorators)

- [Mobx decorators](https://github.com/farwayer/mobx-decorators)
"
`;
